File: packages/health_connector_hc_android/android/src/test/kotlin/com/phamtunglam/health_connector_hc_android/unit_tests/mappers/health_record_mappers/NutritionRecordMapperTest.kt
package com.phamtunglam.health_connector_hc_android.unit_tests.mappers.health_record_mappers

import androidx.health.connect.client.records.MealType
import androidx.health.connect.client.records.NutritionRecord
import androidx.health.connect.client.records.metadata.Metadata
import androidx.health.connect.client.units.grams
import androidx.health.connect.client.units.kilocalories
import com.phamtunglam.health_connector_hc_android.mappers.health_record_mappers.toDto
import com.phamtunglam.health_connector_hc_android.mappers.health_record_mappers.toHealthConnect
import com.phamtunglam.health_connector_hc_android.pigeon.DeviceTypeDto
import com.phamtunglam.health_connector_hc_android.pigeon.EnergyDto
import com.phamtunglam.health_connector_hc_android.pigeon.EnergyUnitDto
import com.phamtunglam.health_connector_hc_android.pigeon.HealthDataTypeDto
import com.phamtunglam.health_connector_hc_android.pigeon.MassDto
import com.phamtunglam.health_connector_hc_android.pigeon.MassUnitDto
import com.phamtunglam.health_connector_hc_android.pigeon.MealTypeDto
import com.phamtunglam.health_connector_hc_android.pigeon.MetadataDto
import com.phamtunglam.health_connector_hc_android.pigeon.NutritionRecordDto
import com.phamtunglam.health_connector_hc_android.pigeon.RecordingMethodDto
import io.kotest.matchers.shouldBe
import java.time.Instant
import java.time.ZoneOffset
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test

@DisplayName("NutritionRecordMapper")
class NutritionRecordMapperTest {

    private companion object {
        const val TEST_ID = "nutrition-id"
        const val TEST_START_TIME = 1609459200000L
        const val TEST_END_TIME = 1609462800000L
        const val TEST_FOOD_NAME = "Banana"
        const val TEST_ENERGY_KCAL = 105.0
        const val TEST_PROTEIN_GRAMS = 1.3
        const val TEST_CARBS_GRAMS = 27.0
        val TEST_ZONE_OFFSET: ZoneOffset = ZoneOffset.ofHours(1)
    }

    @Test
    @DisplayName("GIVEN NutritionRecord → WHEN toDto → THEN converts correctly")
    fun whenRecordToDto_thenConvertsCorrectly() {
        // Given
        val record = NutritionRecord(
            startTime = Instant.ofEpochMilli(TEST_START_TIME),
            endTime = Instant.ofEpochMilli(TEST_END_TIME),
            startZoneOffset = TEST_ZONE_OFFSET,
            endZoneOffset = TEST_ZONE_OFFSET,
            metadata = Metadata.manualEntry(),
            name = TEST_FOOD_NAME,
            mealType = MealType.MEAL_TYPE_BREAKFAST,
            energy = TEST_ENERGY_KCAL.kilocalories,
            protein = TEST_PROTEIN_GRAMS.grams,
            totalCarbohydrate = TEST_CARBS_GRAMS.grams,
        )

        // When
        val result = record.toDto()

        // Then
        result.startTime shouldBe TEST_START_TIME
        result.endTime shouldBe TEST_END_TIME
        result.startZoneOffsetSeconds shouldBe TEST_ZONE_OFFSET.totalSeconds.toLong()
        result.endZoneOffsetSeconds shouldBe TEST_ZONE_OFFSET.totalSeconds.toLong()
        result.foodName shouldBe TEST_FOOD_NAME
        result.mealType shouldBe MealTypeDto.BREAKFAST
        result.energy?.value shouldBe TEST_ENERGY_KCAL
        result.energy?.unit shouldBe EnergyUnitDto.KILOCALORIES
        result.protein?.value shouldBe TEST_PROTEIN_GRAMS / 1000.0 // Convert grams to kgs
        result.protein?.unit shouldBe MassUnitDto.KILOGRAMS
        result.totalCarbohydrate?.value shouldBe TEST_CARBS_GRAMS / 1000.0 // Convert grams to kgs
        result.totalCarbohydrate?.unit shouldBe MassUnitDto.KILOGRAMS
        result.healthDataType shouldBe HealthDataTypeDto.NUTRITION
    }

    @Test
    @DisplayName("GIVEN NutritionRecordDto → WHEN toHealthConnect → THEN creates correctly")
    fun whenDtoToHealthConnect_thenCreatesCorrectly() {
        // Given
        val dto = NutritionRecordDto(
            id = TEST_ID,
            startTime = TEST_START_TIME,
            endTime = TEST_END_TIME,
            startZoneOffsetSeconds = TEST_ZONE_OFFSET.totalSeconds.toLong(),
            endZoneOffsetSeconds = TEST_ZONE_OFFSET.totalSeconds.toLong(),
            healthDataType = HealthDataTypeDto.NUTRITION,
            foodName = TEST_FOOD_NAME,
            mealType = MealTypeDto.LUNCH,
            energy = EnergyDto(value = TEST_ENERGY_KCAL, unit = EnergyUnitDto.KILOCALORIES),
            protein = MassDto(value = TEST_PROTEIN_GRAMS, unit = MassUnitDto.GRAMS),
            metadata = MetadataDto(
                dataOrigin = "com.example.app",
                recordingMethod = RecordingMethodDto.MANUAL_ENTRY,
                deviceType = DeviceTypeDto.PHONE,
            ),
        )

        // When
        val result = dto.toHealthConnect()

        // Then
        result.startTime shouldBe Instant.ofEpochMilli(TEST_START_TIME)
        result.endTime shouldBe Instant.ofEpochMilli(TEST_END_TIME)
        result.startZoneOffset shouldBe TEST_ZONE_OFFSET
        result.endZoneOffset shouldBe TEST_ZONE_OFFSET
        result.name shouldBe TEST_FOOD_NAME
        result.mealType shouldBe MealType.MEAL_TYPE_LUNCH
        result.energy?.inKilocalories shouldBe TEST_ENERGY_KCAL
        result.protein?.inGrams shouldBe TEST_PROTEIN_GRAMS
        result.metadata.id shouldBe TEST_ID
    }
}

File: packages/health_connector_hk_ios/test/unit_tests/src/mappers/health_record_mappers/nutrition/nutrition_record_mapper_test.dart
import 'package:health_connector_core/health_connector_core_internal.dart';
import 'package:health_connector_hk_ios/src/mappers/health_record_mappers/nutrition/nutrition_record_mapper.dart';
import 'package:health_connector_hk_ios/src/pigeon/health_connector_hk_ios_api.g.dart';
import 'package:test/test.dart';

import '../../../../utils/fake_data.dart';

void main() {
  group(
    'NutritionRecordMapper',
    () {
      group(
        'NutritionRecordToDto',
        () {
          test(
            'converts NutritionRecord to NutritionRecordDto',
            () {
              final record = NutritionRecord(
                id: HealthRecordId(FakeData.fakeId),
                startTime: FakeData.fakeStartTime,
                endTime: FakeData.fakeEndTime,
                startZoneOffsetSeconds: FakeData.fakeStartTimeZoneOffsetSeconds,
                endZoneOffsetSeconds: FakeData.fakeEndTimeZoneOffsetSeconds,
                metadata: const Metadata(
                  dataOrigin: DataOrigin(FakeData.fakeDataOrigin),
                  recordingMethod: RecordingMethod.manualEntry,
                  clientRecordVersion: 1,
                  device: Device(type: DeviceType.phone),
                ),
                mealType: MealType.lunch,
                foodName: 'Chicken Salad',
                energy: const Energy.calories(500),
                protein: const Mass.grams(30),
                totalCarbohydrate: const Mass.grams(40),
                totalFat: const Mass.grams(20),
              );

              final dto = record.toDto();

              expect(dto.id, FakeData.fakeId);
              expect(
                dto.startTime,
                FakeData.fakeStartTime.millisecondsSinceEpoch,
              );
              expect(dto.endTime, FakeData.fakeEndTime.millisecondsSinceEpoch);
              expect(dto.mealType, MealTypeDto.lunch);
              expect(dto.foodName, 'Chicken Salad');
              expect(dto.energy?.kilocalories, 0.5);
              expect(dto.protein?.kilograms, 0.03);
              expect(dto.metadata.dataOrigin, FakeData.fakeDataOrigin);
            },
          );
        },
      );

      group(
        'NutritionRecordDtoToDomain',
        () {
          test(
            'converts NutritionRecordDto to NutritionRecord',
            () {
              final dto = NutritionRecordDto(
                id: FakeData.fakeId,
                startTime: FakeData.fakeStartTime.millisecondsSinceEpoch,
                endTime: FakeData.fakeEndTime.millisecondsSinceEpoch,
                startZoneOffsetSeconds: FakeData.fakeStartTimeZoneOffsetSeconds,
                endZoneOffsetSeconds: FakeData.fakeEndTimeZoneOffsetSeconds,
                metadata: MetadataDto(
                  dataOrigin: FakeData.fakeDataOrigin,
                  recordingMethod: RecordingMethodDto.activelyRecorded,
                  clientRecordVersion: 1,
                  deviceType: DeviceTypeDto.phone,
                ),
                mealType: MealTypeDto.dinner,
                foodName: 'Pasta',
                energy: EnergyDto(kilocalories: 600),
                protein: MassDto(kilograms: 25),
                totalCarbohydrate: MassDto(kilograms: 80),
                totalFat: MassDto(kilograms: 15),
              );

              final record = dto.toDomain();

              expect(record.id.value, FakeData.fakeId);
              expect(record.startTime, FakeData.fakeStartTime);
              expect(record.endTime, FakeData.fakeEndTime);
              expect(record.mealType, MealType.dinner);
              expect(record.foodName, 'Pasta');
              expect(record.energy?.inKilocalories, 600);
              expect(record.protein?.inKilograms, 25);
              expect(
                record.metadata.dataOrigin.packageName,
                FakeData.fakeDataOrigin,
              );
            },
          );

          test(
            'converts NutritionRecordDto with null id to domain with none id',
            () {
              final dto = NutritionRecordDto(
                startTime: FakeData.fakeStartTime.millisecondsSinceEpoch,
                endTime: FakeData.fakeEndTime.millisecondsSinceEpoch,
                startZoneOffsetSeconds: FakeData.fakeStartTimeZoneOffsetSeconds,
                endZoneOffsetSeconds: FakeData.fakeEndTimeZoneOffsetSeconds,
                metadata: MetadataDto(
                  dataOrigin: FakeData.fakeDataOrigin,
                  recordingMethod: RecordingMethodDto.manualEntry,
                  clientRecordVersion: 1,
                  deviceType: DeviceTypeDto.phone,
                ),
                mealType: MealTypeDto.snack,
                foodName: 'Apple',
              );

              final record = dto.toDomain();

              expect(record.id, HealthRecordId.none);
            },
          );
        },
      );
    },
  );
}
